lib/ash/type/map.ex

```diff
@@ -16,26 +74,77 @@ defmodule Ash.Type.Map do
 
   def cast_input(nil, _), do: {:ok, nil}
 
-  def cast_input(value, _) when is_binary(value) do
+  def cast_input(value, constraints) when is_binary(value) do
     case Jason.decode(value) do
       {:ok, value} ->
-        {:ok, value}
+        cast_input(value, constraints)
 
       _ ->
         :error
     end
   end
 
-  def cast_input(value, _) when is_map(value), do: {:ok, value}
+  def cast_input(value, constraints) when is_map(value), do: check_constraints(value, constraints)
   def cast_input(_, _), do: :error
 
   @impl true
   def cast_stored(nil, _), do: {:ok, nil}
+
   def cast_stored(value, _) when is_map(value), do: {:ok, value}
+
   def cast_stored(_, _), do: :error
 
   @impl true
   def dump_to_native(nil, _), do: {:ok, nil}
   def dump_to_native(value, _) when is_map(value), do: {:ok, value}
   def dump_to_native(_, _), do: :error
+
+  defp check_constraints(value, constraints) do
+    Enum.reduce(constraints, {:ok, value}, fn
+      {:fields, fields}, {:ok, value} ->
+        check_fields(value, fields)
+
+      {_, _}, {:error, errors} ->
+        {:error, errors}
+    end)
+  end
+
+  defp check_fields(value, fields) do
+    Enum.reduce(fields, {:ok, %{}}, fn
+      {field, field_constraints}, {:ok, checked_value} ->
+        case fetch_field(value, field) do
+          nil ->
+            if field_constraints[:allow_nil?] == false do
+              {:error, [[message: "must not be nil", field: field]]}
+            else
+              {:ok, Map.put(checked_value, field, nil)}
```

Comment:
Not completely sure about this, but the change overall would lead to all the keys in the constraints always being present (set to nil) and extra keys being removed. 

I think I prefer it this way.

Reply:
If you change `fetch_field` to return `:error` when neither key is present then you can detect the two cases. And not add keys to the map that were missing in the input.

Reply:
Do you know how Absinthe will react if we create the graphql type from the constraints and the map that gets passed does not contain all the keys of the type?